### ISSUE TITLE
allow sending multiple requests at once

### ISSUE THREAD

**Issue by patrick-zippenfenig** [vapor/redis#111]
Running map(FutureA, FutureB) or [Future<T>].flatten() now triggers:
````swift
assert(currentSend == nil, "Attempting to call `send(...)` again before previous invocation has completed.")
````
Change was introduced here: https://github.com/vapor/redis/commit/883435da0a32aa212d5aa109904115def42ad41f#diff-3ad0999dab777cc805809e2bb88343ddR59

Is this an intended behaviour?

**Comment by patrick-zippenfenig** [comment]
As a workaround I now use a lazy synchronised flaten:

```swift
public typealias LazyFuture<T> = () -> (Future<T>)
func syncLazyFlatten<T>(futures: [LazyFuture<T>], on worker: Worker) -> Future<[T]> {
    let promise = worker.eventLoop.newPromise([T].self)
    var results = [T]()
    var iterator = futures.makeIterator()
    func handle(_ future: LazyFuture<T>) {
        future().do { result in
            results.append(result)
            if let next = iterator.next() {
                handle(next)
            } else {
                promise.succeed(result: results)
            }
        }.catch { e in
            promise.fail(error: e)
        }
    }
    if let first = iterator.next() {
        handle(first)
    } else {
        promise.succeed(result: results)
    }
    return promise.futureResult
}
```

**Comment by tanner0101** [comment]
Async module also has `syncFlatten` method: https://api.vapor.codes/core/latest/Async/Extensions/Collection.html#/s:s10CollectionP5Async3NIO15EventLoopFutureCyytGyc7ElementRtzrlE11syncFlattenAgD0dE5Group_p2on_tF

I'd be interested to see the application code you wrote that is triggering this error.

**Comment by patrick-zippenfenig** [comment]
I run multiple redis commands and wait until all data is available. ``syncFlatten`` does not return a result. Otherwise it would work.

Small test case:
````swift
let container = try Application(...)
let data = try container.withPooledConnection(to: .redis) { redis -> Future<Void> in
    let foo = redis.get("foo", as: String.self)
    let bar = redis.get("bar", as: String.self)
    return map(to: Void.self, foo, bar) { f, b in
        return
    }
}.wait()
````

**Comment by vzsg** [comment]
Redis has a page dedicated [in their documentation](https://redis.io/topics/pipelining) about the actual benefits of "real" pipelining and it should be pretty convincing.

But I think that even without supporting that, any user would reasonably expect that they can start off multiple commands at the same time, and get responses back in the correct order too – without any crazy tricks like using LazyFutures or a chain of flatMaps.

**Comment by EricWVGG** [comment]
I encountered this issue while setting up Redis. I'm not processing multiple futures in tandem, so unsure how I'm crashing into this issue.
`
    func getOneHandler(_ req: Request) throws -> Future<Article> {
        guard let slug = try? req.parameters.next(String.self) else { throw Abort(.notFound) }
        return req.withNewConnection(to: .redis) { redis -> Future<Article>  in
            return redis.get("article-\(slug)", as: Data.self).flatMap { cached -> Future<Article> in
                if cached != nil {
                    let jsonDecoder = JSONDecoder()
                    let article = try jsonDecoder.decode(Article.self, from: cached!)
                    return req.future(article)
                }
                return Article.query(on: req).filter(\.slug == slug).first().flatMap { article -> Future<Article> in
                    guard let article = article else { throw Abort(.notFound) }
                    let jsonEncoder = JSONEncoder()
                    let jsonData = try jsonEncoder.encode(article)
                    return redis.set("article-\(slug)", to: jsonData).map { _ -> Article in
                        return article
                    }
                }
            }
        }
    }
`

**Comment by EricWVGG** [comment]
I re-wrote this controller to use keyedCache and am getting no complaints. Now unsure if my problem is related to this  at all.

fixed code:
```swift
    func getOneHandler(_ req: Request) throws -> Future<Article> {
        guard let slug = try? req.parameters.next(String.self) else { throw Abort(.notFound) }
        return try req.keyedCache(for: .redis).get("article-\(slug)", as: Data.self).flatMap { cached -> Future<Article> in
            if cached != nil {
                let article = try JSONDecoder().decode(Article.self, from: cached!)
                if article.status == .draft { throw Abort(.forbidden) }
                return req.future(article)
            }
            return Article.query(on: req).filter(\.slug == slug).first().flatMap { article -> Future<Response> in
                guard let article = article else { throw Abort(.notFound) }
                if article.status == .draft { throw Abort(.forbidden) }
                let jsonData = try JSONEncoder().encode(article)
                return try req.keyedCache(for: .redis).set("article-\(slug)", to: jsonData).map { _ -> Response in
                    try response.content.encode(article, as: .json)
                    return response
                }
            }
        }
    }
```

**Comment by Mordil** [comment]
This should be resolved as of #149 and Redis 3.4.0
